### DISCUSSION TITLE
Can I generate .NET Core assembly with a Roslyn running under .NET Framework

### DISCUSSION THREAD

**Discussion by yevgeni-zolotko** [dotnet/roslyn#48273] [Q&A] [answered]
I would like to compile a .NET Core executable by invoking Roslyn APIs. My "compiler" application is built for .NET Framework, and references Roslyn NuGet packages. So far the assemblies generated by Roslyn as a result were targeted for .NET Framework. And I didn't find a switch to change that. Is that possible, or do I need to use .NET Core version of Roslyn to generate .NET Core executables?

**Comment by huoyaoyuan** [comment]
See #49030

**Reply by PathogenDavid** [reply]
Only partially, but that issue is what reminded me of this question.

**Comment by PathogenDavid** [comment, marked as answer]
Not sure if this problem is still relevant to you, but I randomly remembered seeing this question and forgetting to answer it.

I believe the only meaningful* difference between a .NET Core assembly and a .NET Framework assembly is which BCL assemblies they reference. As such there's nothing preventing Roslyn running under .NET Framework from generating .NET Core assemblies. (Visual Studio runs Roslyn under .NET Framework, at least in some scenarios.)

If you just want tools like ILSpy to identify your assembly as .NET Core, you need to add [`TargetFrameworkAttribute`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.versioning.targetframeworkattribute) to your code. (This is normally [done automatically by MSBuild](https://github.com/dotnet/msbuild/blob/c10f34c4e00644e06120cb87f323c330f29a1e89/src/Tasks/Microsoft.Common.CurrentVersion.targets#L3327-L3369).)

In the case of executables .NET Framework is somewhat different in that the assembly is generated as an .exe directly, but in .NET Core the assembly is still emitted as a .dll and the .exe is actually an apphost that launches your assembly.

So if you want to generate .NET Core executables, you need to generate that apphost. When you build with the SDK, this is normally handled by the [`_CreateAppHost` target](https://github.com/dotnet/sdk/blob/ee455863aa6ddf13108bb54b37a5a47fb12fe39e/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L440-L468) which is implemented by the [`CreateAppHost` task](https://github.com/dotnet/sdk/blob/d03a5382ba3821777645f45ab91818c594705b24/src/Tasks/Microsoft.NET.Build.Tasks/CreateAppHost.cs) which in turn is powered by [`Microsoft.NET.HostModel.AppHost.HostWriter`](https://github.com/dotnet/runtime/blob/e0e1919aef7176fb7cabd8a85cad26d80e62728f/src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs) which I *believe* can be consumed via the [`Microsoft.NET.HostModel` NuGet package](https://www.nuget.org/packages/Microsoft.NET.HostModel/).

As far as locating the template for the apphost, that I'm not 100% sure about. You can see the logic of how the SDK does it in the previously linked targets file if you want to try and adapt that to your program. You could also [build it from source](https://github.com/dotnet/runtime/blob/v5.0.0-rc.2.20475.5/src/installer/corehost/corehost.cpp) and carry it with your application. If you just want to hard code a path for testing, on Windows the .NET 5 RC2 apphost is located at `C:\Program Files\dotnet\sdk\5.0.100-rc.2.20479.15\AppHostTemplate\apphost.exe`. (If you're going to try and replicate the logic the SDK uses, you may find it helpful to use [MSBuild Log Viewer](https://msbuildlog.com/) to help you follow things from an actual build.)

(*Obviously newer features might be used in .NET Core assemblies too. Roslyn usually (always?) determines the ability to use these features via [`System.Runtime.CompilerServices.RuntimeFeature`](https://github.com/dotnet/runtime/blob/cb873bc51eb3e21e7cfba12bd3dac25518d85d67/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeFeature.cs) and similar types, so they should just work if you reference the newer BCL.)

**Reply by yevgeni-zolotko** [reply]
@PathogenDavid Thanks for the info, this pretty much answers the question for me.

**Reply by PathogenDavid** [reply]
No problem, glad it helped!
